src/auth/linking/producer.ts

```diff
@@ -0,0 +1,294 @@
+import aes from "keystore-idb/lib/aes/index.js"
+import rsa from "keystore-idb/lib/rsa/index.js"
+import utils from "keystore-idb/lib/utils.js"
+import { KeyUse, SymmAlg, HashAlg, CharSize } from "keystore-idb/lib/types.js"
+
+import * as did from "../../did/index.js"
+import * as ucan from "../../ucan/index.js"
+import { impl as auth } from "../implementation.js"
+import { EventEmitter } from "../../common/event-emitter.js"
+import { LinkingError, LinkingWarning, handleLinkingError, tryParseMessage } from "../linking.js"
+
+import type { Maybe, Result } from "../../common/index.js"
+import type { EventListener } from "../../common/event-emitter.js"
+
+export type AccountLinkingProvider = {
+  on: OnChallenge & OnLink & OnDone
+  cancel: () => void
+}
+
+type OnChallenge = (
+  event: "challenge",
+  listener: (
+    args: {
+      pin: number[]
+      confirmPin: () => void
+      rejectPin: () => void
+    }
+  ) => void
+) => void
+type OnLink = (event: "link", listener: (args: { approved: boolean; username: string }) => void) => void
+type OnDone = (event: "done", listener: () => void) => void
+
+type LinkingStep = "BROADCAST" | "NEGOTIATION" | "DELEGATION"
+
+type LinkingState = {
+  username: Maybe<string>
+  sessionKey: Maybe<CryptoKey>
+  temporaryRsaPair: Maybe<CryptoKeyPair>
+  step: Maybe<LinkingStep>
+}
+
+/**
+ * Create an account linking provider
+ *
+ * @param options provider options
+ * @param options.username username of the account
+ * @returns an account linking event emitter and cancel function
+ */
+export const createProducer = async (options: { username: string }): Promise<AccountLinkingProvider> => {
+  const { username } = options
+  const canDelegate = await auth.checkCapability(username)
+
+  if (!canDelegate) {
+    throw new LinkingError(`Producer cannot delegate for username ${username}`)
+  }
+
+  let eventEmitter: Maybe<EventEmitter> = new EventEmitter()
+  const ls: LinkingState = {
+    username,
+    sessionKey: null,
+    temporaryRsaPair: null,
+    step: "BROADCAST"
+  }
+
+  const handleMessage = async (event: MessageEvent): Promise<void> => {
+    const { data } = event
+    const message = data.arrayBuffer ? new TextDecoder().decode(await data.arrayBuffer()) : data
+
+    if (ls.step === "BROADCAST") {
+      const { sessionKey, sessionKeyMessage } = await generateSessionKey(message)
+      ls.sessionKey = sessionKey
+      ls.step = "NEGOTIATION"
+      channel.send(sessionKeyMessage)
+    } else if (ls.step === "NEGOTIATION") {
+      if (ls.sessionKey) {
+        const userChallengeResult = await handleUserChallenge(ls.sessionKey, message)
+        ls.step = "DELEGATION"
+
+        if (userChallengeResult.ok) {
+          const { pin, audience } = userChallengeResult.value
+
+          const challengeOnce = () => {
+            let called = false
+
+            return {
+              confirmPin: async () => {
+                if (!called) {
+                  called = true
```

Comment:
It's good that we have this check in here! A user might accidentally e.g. call *both* `confirmPin()` and `rejectPin()`.

Because this is a user error, can we throw an error in the case `called` is true?

Reply:
To be clear: I absolutely agree on having the `if (!called)` in there.
I was just thinking, maybe having an `else throw new Error(...)` in there might be nice as well :)

Reply:
Maybe I'm misunderstanding you - is there a case where throwing would be bad?

Reply:
Yes, there is. I may not be explaining it well.

If a provider links for more than one requestor, all `confirmPin` or `rejectPin` functions it has emitted will be called. This would lead to exceptions in the functions that had been marked as called.

I'd be happy to discuss this synchronously and demonstrate where I think this would go wrong. There may be a simple solution that I am overlooking here.